prov/sm2/src/sm2_atomic.c

```diff
@@ -183,15 +182,24 @@ static inline ssize_t sm2_generic_atomic(
 		rma_ioc, rma_ioc_count, compare_iov, compare_count, result_iov,
 		result_count, total_len, context, proto_flags);
 
-	if (ret)
+	if (ret) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL, "Atomic operation failed\n");
 		goto out;
+	}
 
-	if (!(proto_flags & SM2_ATOMIC_RMA_REQ) &&
-	    !(op_flags & FI_DELIVERY_COMPLETE)) {
-		ret = sm2_complete_tx(ep, context, op, op_flags);
-		if (ret)
-			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
-				"unable to process tx completion\n");
+	/* Generate immediate send completion for ofi_op_atomic unless
+	 * FI_DELIVERY_COMPLETE are set */
+	switch (op) {
+	case ofi_op_atomic:
+		if (!(op_flags & FI_DELIVERY_COMPLETE)) {
+			ret = sm2_complete_tx(ep, context, op, op_flags);
```

Comment:
I think we need a different check here. We can't complete yet for read requests

Reply:
Do you mean RMA read? RMA is not in SM2

Reply:
No, I meant an atomic with FI_ATOMIC_READ but I see this is only for the base atomics, and the atomic read is a fetch atomic so nevermind!
Looking at this again though, I think a single if statement here would work better instead of a switch.

Reply:
I can turn it into an `if`